### PR TITLE
[dagster-databricks] databricks_pyspark_step_launcher supports retries

### PR DESCRIPTION
**Pull request by OwenKephart**
Resolves: https://github.com/dagster-io/dagster/issues/7122

Step launchers are weird and hard to reason about.

The basic functionality of a step launcher is to:

a) Serialize all of the relevant state about the host environment (so step context, relevant logs, etc.)
b) Ship that state (and the code you want to run) over to some external execution environment
c) Deserialize that state in the external execution environment to match the original environment as closely as possible
d) Run the code in the external environment, keeping track of any state changes
e) Serialize a representation of those state changes and ship it back to the host process
f) Have the host process ingest those changes and "make it seem like" they happened natively in the host process

Prior to this PR, the "keeping track of any state changes" in d) was done by simply recording all of the entries that were added to the event log on the external instance. However, it's not actually the case that any event which alters the control flow of Dagster will have a corresponding representation in the event log. In particular, there are two Dagster-specific exceptions which have special properties in terms of determining how Dagster will run a particular step: `RetryRequested` and `Failure`.

When you set a RetryPolicy on an op, what happens is that the process that executes that op will capture any exception and, depending on the policy, raise a RetryRequested exception telling the dagster framework to retry the step, and how long to wait before the next retry. These exceptions would not get captured / sent over to the host process, meaning that the policy would not be respected. A similar (but less impactful) issue would occur with Failure exceptions.

This PR fixes the underlying issue by exporting these serialized exceptions back to the host process so that they can be raised again.

This doesn't preserve the stack trace of the original exception, but it will at least manage the control flow properly.

### CODE REVIEW

File: python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py [hunk 1]
```diff
@@ -240,8 +244,9 @@ def step_events_iterator(self, step_context, step_key: str, databricks_run_id: i
 
         step_context.log.info(f"Databricks run {databricks_run_id} completed.")
 
-    def get_step_events(self, run_id: str, step_key: str):
-        path = self._dbfs_path(run_id, step_key, PICKLED_EVENTS_FILE_NAME)
+    def get_step_events(self, run_id: str, step_key: str, retry_number: int):
+        # events for each retry written to a different file
+        path = self._dbfs_path(run_id, step_key, f"{retry_number}{PICKLED_EVENTS_FILE_NAME}")
```
Comment (alangenfeld) [inline python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py:249]: delimiter between retry # and file name? an aesthetic choice in the end

**Review by alangenfeld — COMMENTED** [review]
Good work digging through the mud on this. 

Are there any other step launchers that need updating? 

Would capturing the exception object and re-throwing be a better UX for all exception types?

**Comment by alangenfeld** [comment]
> This doesn't preserve the stack trace of the original exception, but it will at least manage the control flow properly. 

ah missed this - whats the context here?

**Comment by OwenKephart** [comment]
@alangenfeld 

Re: "Any other step launchers that need updating?" -- The pyspark emr step launcher will need to be updated separately (I believe there are actually some other fixes in this vein that it's missed out on as well). I can put aside some time to do that (it'll probably be a larger diff than this one).

Re: "Would capturing the exception object and re-throwing be a better UX for all exception types?" (and also the question about tracebacks) -- This was something I really wanted to get working properly, but it runs into some issues. First, you'll basically only ever catch one of three exception types with this code: RetryRequested, Failure, or DagsterUserCodeExecutionError (because this is used to wrap stuff that comes out of the user code error boundary). RetryRequested and Failure are picklable, but DagsterUser...Error is not (both because it contains unpicklable traceback stuff and also because the *args/**kwargs types of constructors are not very portable without defining an explict `__reduce__` function). One option I considered was to forget about pickling that entire class, and instead just pickle the underlying `DagsterUser...Error.user_exception` exception, but a) this is guaranteed to be possible to pickle (although it is much more likely) and b)  regardless of if we can get the error back into the host process or not, an exception will not contain a traceback, so raising it in the host process will just give the traceback that the host process followed to raise the exception (which will just be `raise pickle.load(pickled_exception)`).

However..... we do have a serialized_error_info_from_exc_info, which might be possible to hack into a solution

**Review by alangenfeld — APPROVED** [review]
oh `pickle`. 


Can we record this snapshot of understanding on to the step launcher base class or something? we have this pr discussion I think its worth putting something a little more condensed in to the code base as well.

**Comment by OwenKephart** [comment]
@alangenfeld After agonizing about how unaesthetic this change was, I figured out a cleaner implementation: https://github.com/dagster-io/dagster/pull/7171

**Pull request closed without merging**